apps/src/music/progress/ProgressManager.ts

```diff
@@ -0,0 +1,129 @@
+// This file contains a generic ProgressManager which any lab can include,
+// if it wants to make progress without reloading the page.
+
+// Abstract class that validates a set of conditions. How
+// the validation works is up to the implementor.
+export abstract class Validator {
+  abstract shouldCheckConditions(): boolean;
+  abstract checkConditions(): void;
+  abstract conditionsMet(conditions: string[]): boolean;
+  abstract clear(): void;
+}
+
+// A validation inside the progression.
+interface Validation {
+  conditions: string[];
+  message: string;
+  next: boolean;
+}
+
+// The definition of the progression.
+interface Progression extends Object {
+  panels: {
+    text: string;
+    toolbox: {
+      [key: string]: string;
+    };
+    sounds: {
+      [key: string]: string;
+    };
+    validations: Validation[];
+  }[];
+}
+
+// The current progress state.
+interface ProgressState {
+  step: number;
+  satisfied: boolean;
+  message?: string | null;
+}
+
+export default class ProgressManager {
+  private progression: Progression;
+  private validator: Validator;
+  private onProgressChange: (state: ProgressState) => void;
+  private currentProgressState: ProgressState;
+
+  constructor(
+    progression: Progression,
+    validator: Validator,
+    onProgressChange: (state: ProgressState) => void
+  ) {
+    this.progression = progression;
+    this.validator = validator;
+    this.onProgressChange = onProgressChange;
+    this.currentProgressState = {
+      step: 0,
+      satisfied: false,
+      message: null
+    };
+  }
+
+  getProgression() {
+    return this.progression;
+  }
+
+  getCurrentState() {
+    return this.currentProgressState;
+  }
+
+  getCurrentToolbox() {
+    return this.progression.panels[this.currentProgressState.step].toolbox;
+  }
+
+  getCurrentSounds() {
+    return this.progression.panels[this.currentProgressState.step].sounds;
+  }
+
+  updateProgress(): void {
+    const validations = this.progression.panels[this.currentProgressState.step]
+      .validations;
+
+    if (!validations) {
+      return;
+    }
+
+    // Find out from the lab-specific code whether we should be trying to
+    // check conditions at the moment.  Otherwise, we might get a fail
+    // when we shouldn't have even been checking.
+    if (!this.validator.shouldCheckConditions()) {
```

Comment:
Curious about this - if the lab itself is calling `updateProgress`, can we trust that the lab will only call it when it's appropriate to check conditions?

Reply:
It's a great question, and something to keep in mind as Music Lab gets more states, especially for things like loading data, transitioning levels, etc.  We could already add a bit of nuance to Music Lab, so that it only calls this when playing, for example, but I think what we have is alright for a first pass.

It seems that if `ProgressManager` is to remain lab agnostic, we need to ultimately place that responsibility on the caller.  I guess we could add some more states to `ProgressManager`, allowing the caller to give it some context as to when to check state, but I'm not quite sure what they would be yet.